File: backend/src/routes/jobs.ts
import { Router } from 'express';
import { JobModel } from '../models/Job.ts';
import { authenticateToken } from '../middleware/auth.ts';
import { fetchJobs as fetchPublicJobs } from '../../services/api.ts';

const router = Router();
let jobModel: JobModel;

export function setJobModel(db: any) {
  jobModel = JobModel.create(db);
}

const serializeTags = (tags: any) => {
  if (Array.isArray(tags)) return JSON.stringify(tags.slice(0, 8));
  if (tags === undefined || tags === null) return undefined;
  return String(tags);
};

const toNumber = (value: any, fallback?: number) => {
  const numericValue = Number(value);
  return Number.isFinite(numericValue) ? numericValue : fallback;
};

const syncPublicJobs = async ({
  search = '',
  category = '',
  limit = 40,
}: {
  search?: string;
  category?: string;
  limit?: number;
}) => {
  try {
    const publicJobs = await fetchPublicJobs({
      search,
      category,
      limit: Math.min(Math.max(limit, 1), 80),
    });

    for (const publicJob of publicJobs) {
      const externalId = publicJob.id ? String(publicJob.id) : '';
      if (!externalId) continue;

      const existing = await jobModel.findByExternalId(externalId);
      const jobPayload = {
        externalId,
        source: 'public',
        title: publicJob.title,
        company: publicJob.company,
        companyLogo: publicJob.companyLogo,
        category: publicJob.category,
        rawCategory: publicJob.rawCategory,
        jobType: publicJob.jobType,
        location: publicJob.location,
        salary: publicJob.salary,
        description: publicJob.description,
        fullDescription: publicJob.fullDescription,
        url: publicJob.url,
        applicationUrl: publicJob.applicationUrl,
        applyUrl: publicJob.applyUrl,
        tags: serializeTags(publicJob.tags),
        postedAt: publicJob.postedAt,
        sourceName: publicJob.source || 'Remotive',
        featured: publicJob.featured || false,
      };

      if (existing?.id) {
        await jobModel.update(existing.id, jobPayload);
      } else {
        await jobModel.create(jobPayload);
      }
    }
  } catch (error: any) {
    console.warn('Public job sync skipped:', error?.message || error);
  }
};

router.get('/', async (req, res) => {
  try {
    const { search, category, limit, offset, sync } = req.query;
    const numericLimit = toNumber(limit, 40) || 40;
    const searchValue = (search as string) || '';
    const categoryValue = (category as string) || '';

    if (sync !== 'false') {
      await syncPublicJobs({
        search: searchValue,
        category: categoryValue,
        limit: numericLimit,
      });
    }

    const jobs = await jobModel.list({
      search: searchValue || undefined,
      category: categoryValue || undefined,
      limit: numericLimit,
      offset: toNumber(offset),
    });

    res.json({ data: jobs });
  } catch (error) {
    console.error('Get jobs error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/my/jobs', authenticateToken, async (req, res) => {
  try {
    const jobs = await jobModel.list({
      postedByUserId: req.user!.id
    });
    res.json({ data: jobs });
  } catch (error) {
    console.error('Get my jobs error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/:id', async (req, res) => {
  try {
    const id = Number(req.params.id);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid job ID' });
    }

    const job = await jobModel.findById(id);
    if (!job) {
      return res.status(404).json({ error: 'Job not found' });
    }

    res.json({ data: job });
  } catch (error) {
    console.error('Get job by ID error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/', authenticateToken, async (req, res) => {
  try {
    const {
      externalId, source, title, company, companyLogo, category, rawCategory,
      jobType, location, salary, description, fullDescription, url,
      applicationUrl, applyUrl, tags, postedAt, sourceName, featured
    } = req.body;

    if (!title || !company) {
      return res.status(400).json({ error: 'Title and company are required' });
    }

    const job = await jobModel.create({
      externalId,
      source: source || 'internal',
      title,
      company,
      companyLogo,
      category,
      rawCategory,
      jobType,
      location,
      salary,
      description,
      fullDescription: fullDescription || description,
      url,
      applicationUrl,
      applyUrl,
      tags: serializeTags(tags),
      postedAt,
      sourceName: sourceName || 'HireFlow',
      featured: featured || false,
      postedByUserId: req.user!.id
    });

    res.status(201).json({ data: job });
  } catch (error) {
    console.error('Create job error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.put('/:id', authenticateToken, async (req, res) => {
  try {
    const id = Number(req.params.id);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid job ID' });
    }

    const job = await jobModel.findById(id);
    if (!job) {
      return res.status(404).json({ error: 'Job not found' });
    }

    const isOwner = job.postedByUserId === req.user!.id;
    const isAdmin = req.user!.role === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({ error: 'Insufficient permissions' });
    }

    const {
      externalId, source, title, company, companyLogo, category, rawCategory,
      jobType, location, salary, description, fullDescription, url,
      applicationUrl, applyUrl, tags, postedAt, sourceName, featured
    } = req.body;

    const updatedJob = await jobModel.update(id, {
      externalId,
      source,
      title,
      company,
      companyLogo,
      category,
      rawCategory,
      jobType,
      location,
      salary,
      description,
      fullDescription,
      url,
      applicationUrl,
      applyUrl,
      tags: serializeTags(tags),
      postedAt,
      sourceName,
      featured
    });

    if (!updatedJob) {
      return res.status(404).json({ error: 'Job not found' });
    }

    res.json({ data: updatedJob });
  } catch (error) {
    console.error('Update job error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.delete('/:id', authenticateToken, async (req, res) => {
  try {
    const id = Number(req.params.id);
    if (isNaN(id)) {
      return res.status(400).json({ error: 'Invalid job ID' });
    }

    const job = await jobModel.findById(id);
    if (!job) {
      return res.status(404).json({ error: 'Job not found' });
    }

    const isOwner = job.postedByUserId === req.user!.id;
    const isAdmin = req.user!.role === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({ error: 'Insufficient permissions' });
    }

    const deleted = await jobModel.delete(id);
    if (!deleted) {
      return res.status(404).json({ error: 'Job not found' });
    }

    res.json({ message: 'Job deleted successfully' });
  } catch (error) {
    console.error('Delete job error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/src/routes/jobs.py
"""
Jobs Routes Handle job-related API endpoints 
"""
from flask import request
from extensions import db
from models import Job
from utils.auth import authenticate_token, authorize_role
from utils.responses import success_response, error_response, paginated_response
from utils import serialize_tags, validate_string


def register_job_routes(app):
    @app.route('/api/jobs', methods=['GET'])
    def list_jobs():
        try:
            search = request.args.get('search', '').strip()
            category = request.args.get('category', '').strip()
            limit = min(int(request.args.get('limit', 40)), 80)
            offset = int(request.args.get('offset', 0))

            if limit < 1 or offset < 0:
                return error_response('Invalid limit or offset', 400)

            query = Job.query
            if search:
                search_filter = f'%{search}%'
                query = query.filter(
                    db.or_(
                        Job.title.ilike(search_filter),
                        Job.company.ilike(search_filter),
                        Job.description.ilike(search_filter)
                    )
                )
            if category:
                category_filter = f'%{category}%'
                query = query.filter(Job.category.ilike(category_filter))

            total = query.count()
            jobs = query.order_by(Job.posted_at.desc(), Job.created_at.desc()).limit(limit).offset(offset).all()
            jobs_data = [
                {
                    'id': j.id,
                    'externalId': j.external_id,
                    'source': j.source,
                    'title': j.title,
                    'company': j.company,
                    'companyLogo': j.company_logo,
                    'category': j.category,
                    'rawCategory': j.raw_category,
                    'jobType': j.job_type,
                    'location': j.location,
                    'salary': j.salary,
                    'description': j.description,
                    'fullDescription': j.full_description,
                    'url': j.url,
                    'applicationUrl': j.application_url,
                    'applyUrl': j.apply_url,
                    'tags': j.tags,
                    'postedAt': j.posted_at.isoformat() if j.posted_at else None,
                    'sourceName': j.source_name,
                    'featured': j.featured,
                    'postedByUserId': j.posted_by_user_id,
                    'createdAt': j.created_at.isoformat(),
                    'updatedAt': j.updated_at.isoformat()
                }
                for j in jobs
            ]
            return paginated_response(jobs_data, total, limit, offset)
        except ValueError:
            return error_response('Invalid pagination parameters', 400)
        except Exception as e:
            return error_response(f'Error fetching jobs: {str(e)}', 500)

    @app.route('/api/jobs/<int:job_id>', methods=['GET'])
    def get_job(job_id):
        job = Job.query.get(job_id)
        if not job:
            return error_response('Job not found', 404)
        return success_response({
            'id': job.id,
            'externalId': job.external_id,
            'source': job.source,
            'title': job.title,
            'company': job.company,
            'companyLogo': job.company_logo,
            'category': job.category,
            'rawCategory': job.raw_category,
            'jobType': job.job_type,
            'location': job.location,
            'salary': job.salary,
            'description': job.description,
            'fullDescription': job.full_description,
            'url': job.url,
            'applicationUrl': job.application_url,
            'applyUrl': job.apply_url,
            'tags': job.tags,
            'postedAt': job.posted_at.isoformat() if job.posted_at else None,
            'sourceName': job.source_name,
            'featured': job.featured,
            'postedByUserId': job.posted_by_user_id,
            'createdAt': job.created_at.isoformat(),
            'updatedAt': job.updated_at.isoformat()
        })

    @app.route('/api/jobs', methods=['POST'])
    @authenticate_token
    @authorize_role('employer', 'admin')
    def create_job():
        data = request.get_json() or {}
        title = data.get('title', '').strip()
        company = data.get('company', '').strip()

        if not validate_string(title, min_length=3, max_length=255):
            return error_response('Job title must be 3-255 characters', 400)
        if not validate_string(company, min_length=1, max_length=255):
            return error_response('Company name required', 400)

        job = Job(
            source='internal',
            title=title,
            company=company,
            company_logo=data.get('companyLogo', '').strip() or None,
            category=data.get('category', '').strip() or None,
            job_type=data.get('jobType', '').strip() or None,
            location=data.get('location', '').strip() or None,
            salary=data.get('salary', '').strip() or None,
            description=data.get('description', '').strip() or None,
            full_description=data.get('fullDescription', '').strip() or None,
            url=data.get('url', '').strip() or None,
            tags=serialize_tags(data.get('tags')),
            source_name='HireFlow',
            featured=bool(data.get('featured', False)),
            posted_by_user_id=request.user.get('id')
        )
        db.session.add(job)
        db.session.commit()
        return success_response({
            'id': job.id,
            'title': job.title,
            'company': job.company,
            'location': job.location,
            'createdAt': job.created_at.isoformat()
        }, 'Job posted successfully', 201)

    @app.route('/api/jobs/<int:job_id>', methods=['PUT'])
    @authenticate_token
    def update_job(job_id):
        job = Job.query.get(job_id)
        if not job:
            return error_response('Job not found', 404)
        if job.posted_by_user_id != request.user.get('id') and request.user.get('role') != 'admin':
            return error_response('Insufficient permissions', 403)

        data = request.get_json() or {}
        if 'title' in data:
            title = data['title'].strip()
            if not validate_string(title, min_length=3, max_length=255):
                return error_response('Job title must be 3-255 characters', 400)
            job.title = title
        if 'description' in data:
            job.description = data['description'].strip() or None
        if 'fullDescription' in data:
            job.full_description = data['fullDescription'].strip() or None
        if 'salary' in data:
            job.salary = data['salary'].strip() or None
        if 'location' in data:
            job.location = data['location'].strip() or None
        if 'tags' in data:
            job.tags = serialize_tags(data['tags'])
        if 'featured' in data:
            job.featured = bool(data['featured'])

        db.session.commit()
        return success_response(None, 'Job updated successfully')

    @app.route('/api/jobs/<int:job_id>', methods=['DELETE'])
    @authenticate_token
    def delete_job(job_id):
        job = Job.query.get(job_id)
        if not job:
            return error_response('Job not found', 404)
        if job.posted_by_user_id != request.user.get('id') and request.user.get('role') != 'admin':
            return error_response('Insufficient permissions', 403)

        db.session.delete(job)
        db.session.commit()
        return success_response(None, 'Job deleted successfully')
